### ISSUE TITLE
Minimising mechanistic variability

### ISSUE THREAD

**Issue by charlesdgburns** [charlesdgburns/NM_TinyRNN#17]
Kris suggested that some of the variability in mechanisms across models may be due to the fact that GRUs and LSTMs are too expressive for a simple AB reversal learning task. This seems very true.

We should design architectures aligned with our aim of explaining a 1 dimensional neuromodulation signal (or two 1D neuromodulation signals).

**Other minimal GRUs**
- minimal GRU throws away the reset gate and just has an update gate:
- light GRU, notably uses RELU activation after batch normalisation of inputs.

**monoGRU**
Here we want a 1D gating signal that globally modulates the input

$$ h_t = z_th_{t-1} +(1-z_t) n_t $$

where $z_t = \sigma(W_{iz} x_{t-1}+W_{hz} h_{t-1}+b_z)$ and $n_t = \phi(W_{ih} x_{t-1} + W_{hh} h_{t-1}+b_h)$.
This is theoretically related to a leaky integrator network with a dynamically modulated 'global state-update rate' [(Livi et al., 2025)](https://arxiv.org/pdf/2508.12121).

Note that traditional GRU architectures have a gate / state-update-rate per unit, which adds a bit of (unnecessary) complexity and goes against our interpretability aims.

**stereoGRU**

The idea here is to have separate gates for the recurrent (within network) activity and input (external) activity.

$$h_t = \phi \left[ W_{ih} (i_t \cdot x_{t-1}) + W_{hh} (r_t\cdot h_{t-1}) + b_h \right] $$

Where the input and reset gates are again 1D.

NOTE:
In both the mono-gated and stereo-gated RNNs there's a freedom of choice in how the 1D signal is generated.
- access to hidden state versus subnetwork state; we call the variant with a subnetwork state monoGRU2
- sigmoid or tanh activation

Arguably the most simple (and therefore preferred) would be gates of the form:

$$ \text{gate}_t = \sigma(W_{ig} x_{t-1} + W_{hg} h_{t-1}+b_g) $$

Furthermore, underlying mechanisms might also be shaped by inputs (unipolar versus bipolar, see #15) or constraints on weight sparsity versus energy (hidden state activations), see comment below.

**Comment by charlesdgburns** [comment]
**Is it really the decoder initialisation?**

Looking further into it, this might not be the case. 
The difference in performance with and without decoder initialisation was previously driven by *poor initialisation of the decoder*.

In particular, the decoder should not be initialised with small weights but random larger rates, satisfied by PyTorch default uniform(-sqrt(n_units),sqrt(n_units)).

When we only initialise rnn parameters as small, then we can get similar mechaisms with weight_seed_1 even with a large learning rate and max 10 epochs:
(NONE OF THE INITIALISATION TRICKS APART FROM SMALLER WEIGHTS ON RNN; also different dataset splitting.)
<img width="1489" height="390" alt="Image" src="https://github.com/user-attachments/assets/f331c2ea-9873-4479-9c86-b50a034de8e2" />

Arguably the best thing to do here is just letting each weight_seed train for as long as required, since this means we get less variability in the validation loss. Can very quickly do a few runs with just 10 epochs max, but with more weight initialisations. 
This kind of favours some initial weights that are inclined to learn faster as well (or are initialised closer to the optimal weights).
The intuition, however, is that letting it run for longer will improve the fit, which seems important. So that's the plan.

**Comment by charlesdgburns** [comment]
### Update:

Here we trained many models varying only the training seed and weight initialisation seed (20 of each). 
Small weights (-1e-3, 1e3) and small learning rate (1e-4). The sparsity lamba is also fixed at 1e-5 and when trained with energy constraints, the energy lamba is fixed at 1e-2.

Since we're interested in possible variability in mechanisms, we focus on 2 unit models here. 

We compare a standard 2-unit-GRU (the Li et al. implementation, so tanh + sparsity constrained) to the monoGRU trained with biological constraints (RELU + sparsity + energy constraints). 

Note that here the monoGRU gate replaces the $x_{t-1}$ vector (past action, past reward) with only the past reward, $r_{t-1}\in \set{0,1} $, which is then mixed with past hidden state. The motivation/idea here is to reduce choice-gated mechanisms, the intuition being that when predictions are good, previous action ($a_{t-1}$) will highly correlate with previous belief ($h_{t-1}$) so the error signal for choice will be larger (and hence bias the gating mechanism). **TODO** We should revisit the full seed search without this change - ideally 'outcome-gating' mechanisms are better behavioural predictors than 'choice-gating' mechanisms.


## Again, variability in performance is primarily due to training seeds
Note also that the monoGRU is slightly more sensitive to weight initialisation, likely due to vanishing gradients of the RELU activation affecting some subset of initial weights. 

<img width="1189" height="290" alt="Image" src="https://github.com/user-attachments/assets/c6893d30-5dc9-40b5-8b5b-dcf5b96cc4e5" />

<img width="1189" height="290" alt="Image" src="https://github.com/user-attachments/assets/f3b056cd-f547-441a-92fb-3c43e417901c" />

Figure 1: top, validation performance as a function of weight seed, points coloured by train seed.
bottom, eval performance as a function of train seed, points coloured by weight seed.


_Takeaway_: we may treat training seed as a hyperparameter and select the seed with the best validation score, before then picking subsequent models' eval performance across training seeds. The eval of the best-performing weight seeds are shown below. Note that, as before, moving from 1 to 2 units gives a marginal benefit. Now with 20 train seeds there's more variability captured in the loss.



<img width="889" height="590" alt="Image" src="https://github.com/user-attachments/assets/568c6800-62cf-492e-acec-8c4d6a6ad24f" />


## 2-unit GRU mechanisms show variability
Showing all 20 train seeds is a lot, but the top 10 performing ones should tell the story

<img width="1575" height="2989" alt="Image" src="https://github.com/user-attachments/assets/473fc985-a9c4-46ce-b5d0-9019dc1295ab" />

## 2-unit monoGRU shows little-to-no variability
Showing all 20 train seeds is a lot, but the top 10 performing ones tell the story. There's just the one (seed 12) which doesn't have an outcome-gating mechanism. High gate value corresponding with high uncertainty. This also predicts higher uncertainty in one side than another (some level of side-bias).

<img width="1595" height="2989" alt="Image" src="https://github.com/user-attachments/assets/7dd16a26-009d-4750-a653-3ac17ea6b370" />

**Comment by charlesdgburns** [comment]
### Similar mechanisms with fewer epochs and without input hacking

Below we train on data from WS18, max 10 epochs, learning rate 1e-2. $W_{iz} \approx [-1,0]$, setting high weight on outcome input, low weight on past choice input. 

Seed 2 (~0.46 eval loss):
<img width="1489" height="390" alt="Image" src="https://github.com/user-attachments/assets/4769d709-f525-4087-aee3-a10ee14cafef" />

Seed 9 (~0.50 eval loss):
<img width="1490" height="390" alt="Image" src="https://github.com/user-attachments/assets/b3680ca3-8f8f-4e50-a034-6b1ea8f4b04a" />


Seed 9 up to 100 epochs (~0.46 eval loss)

<img width="1489" height="390" alt="Image" src="https://github.com/user-attachments/assets/f9f708fa-4d07-4947-9f11-3311e0ab8926" />
